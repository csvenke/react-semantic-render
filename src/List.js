import PropTypes from 'prop-types'

import getRenderMethod from './internal/getRenderMethod'

function List (props) {
  const method = getRenderMethod(props.children, props.render)
  return method ? props.items.map(method) : null
}

List.propTypes = {
  children: PropTypes.func,
  items: PropTypes.arrayOf(PropTypes.any).isRequired,
  render: PropTypes.func
}

export default List
