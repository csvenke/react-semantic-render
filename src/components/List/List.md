### How to use

```jsx static
import React from 'react';
import { List } from 'react-semantics';
```

#### With render prop
##### Output

```jsx
const data = [1,2,3].map(() => faker.name.findName());

<ul>
  <List 
    items={data}
    render={n => (
      <li key={n}>
        <div>{n}</div>
      </li>
    )}
  />
</ul>
```

#### With children
##### Output

```jsx
const data = [1,2,3].map(() => faker.name.findName());

<ul>
  <List items={data}>
    {n => (
      <li key={n}>
        <div>{n}</div>
      </li>
    )}
  </List>
</ul>
```
