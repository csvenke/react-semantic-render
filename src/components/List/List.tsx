import * as PropTypes from 'prop-types';
import * as React from 'react';

export type ListItemCallbackFn = (
  value?: any,
  index?: number,
  array?: any[],
) => React.ReactNode;

export interface IListProps {
  /** Array to map. */
  items: any[];

  /** Shorthand for primary content. */
  render?: ListItemCallbackFn;

  /** Primary content. */
  children?: any;
}

/**
 * Semantic helper component that calls the specified callback function in either `render` or `children` on each element of `items`.
 */
const List: React.SFC<IListProps> = ({ items, render, children }) => {
  const func = children ? children : render;

  if (!func) {
    return null;
  }

  return <React.Fragment>{items.map((a, b, c) => func(a, b, c))}</React.Fragment>;
};

List.propTypes = {
  children: PropTypes.func,
  items: PropTypes.arrayOf(PropTypes.any).isRequired,
  render: PropTypes.func,
};

export default List;
