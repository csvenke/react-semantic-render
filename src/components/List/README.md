### Example usage

```jsx
const data = _.range(10).map(() => faker.name.findName());

const App = () => (
  <ul>
    <List
      items={data}
      render={i => (
        <li key={i}>
          <div>{i}</div>
        </li>
      )}
    />
  </ul>
);

<App />;
```
