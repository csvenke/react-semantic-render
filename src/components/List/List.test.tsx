import * as React from 'react';
import renderer from 'react-test-renderer';

import List from '.';

describe('with render', () => {
  const input = [1, 2, 3];

  test('should return primary content from render', () => {
    const element = <List items={input} render={n => <div key={n}>{n}</div>} />;
    const component = renderer.create(element);
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('with children', () => {
  const input = [1, 2, 3];

  test('should return primary content from children', () => {
    const element = <List items={input}>{n => <div key={n}>{n}</div>}</List>;
    const component = renderer.create(element);
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('with children and render', () => {
  const input = [1, 2, 3];

  test('should return primary content from children and ignore render', () => {
    const element = (
      <List items={input} render={n => <a key={n}>{n}</a>}>
        {n => <div key={n}>{n}</div>}
      </List>
    );
    const component = renderer.create(element);
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('without children and render', () => {
  const input = [1, 2, 3];

  test('should return null', () => {
    const element = <List items={input} />;
    const component = renderer.create(element);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
