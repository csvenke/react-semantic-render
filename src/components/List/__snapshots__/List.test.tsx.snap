// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with children and render should return primary content from children and ignore render 1`] = `
Array [
  <div>
    1
  </div>,
  <div>
    2
  </div>,
  <div>
    3
  </div>,
]
`;

exports[`with children should return primary content from children 1`] = `
Array [
  <div>
    1
  </div>,
  <div>
    2
  </div>,
  <div>
    3
  </div>,
]
`;

exports[`with render should return primary content from render 1`] = `
Array [
  <div>
    1
  </div>,
  <div>
    2
  </div>,
  <div>
    3
  </div>,
]
`;

exports[`without children and render should return null 1`] = `null`;
